src/farm/library.cairo

```diff
@@ -50,15 +78,274 @@ namespace CarbonableFarmer {
+        success: felt
+    ) {
+        // [Check] Current period
+        let (current_time) = get_block_timestamp();
+        let (start) = start_.read();
+        let (period_duration) = period_duration_.read();
+        let end = start + period_duration;
+        with_attr error_message("CarbonableFarmer: No current period") {
+            assert_le(current_time, end);
+        }
+
+        // [Effect] Reset period information
+        start_.write(0);
+        period_duration_.write(0);
+        unlocked_duration_.write(0);
+
+        return (success=TRUE,);
+    }
+
+    func deposite{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        token_id: Uint256
+    ) -> (success: felt) {
+        alloc_locals;
+
+        // [Security] Start reetrancy guard
+        ReentrancyGuard._start();
+
+        // [Check] Locked status
+        let (status) = is_locked();
+        with_attr error_message("CarbonableFarmer: deposites are currently locked") {
+            assert status = FALSE;
+        }
+
+        // [Check] Uint256 compliance
+        with_attr error_message("CarbonableFarmer: token_id is not a valid Uint256") {
+            uint256_check(token_id);
+        }
+
+        // [Interaction] Transfer token_id from caller to contract
+        let (carbonable_project_address) = carbonable_project_address_.read();
+        let (caller) = get_caller_address();
+        let (contract_address) = get_contract_address();
+        IERC721.transferFrom(
+            contract_address=carbonable_project_address,
+            from_=caller,
+            to=contract_address,
+            tokenId=token_id,
+        );
+
+        // [Check] Transfer successful
+        let (owner) = IERC721.ownerOf(
+            contract_address=carbonable_project_address, tokenId=token_id
+        );
+        with_attr error_message("CarbonableFarmer: transfer failed") {
+            assert owner = contract_address;
+        }
+
+        // [Effect] Register the caller with the token id
+        registration_.write(token_id, caller);
+
+        // [Security] End reetrancy guard
+        ReentrancyGuard._end();
+
+        return (success=TRUE,);
+    }
+
+    func withdraw{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        token_id: Uint256
+    ) -> (success: felt) {
+        alloc_locals;
+
+        // [Security] Start reetrancy guard
+        ReentrancyGuard._start();
+
+        // [Check] Locked status
+        let (status) = is_locked();
+        with_attr error_message("CarbonableFarmer: withdrawals are currently locked") {
+            assert status = FALSE;
+        }
+
+        // [Check] Uint256 compliance
+        with_attr error_message("CarbonableFarmer: token_id is not a valid Uint256") {
+            uint256_check(token_id);
+        }
+
+        // [Effect] Remove the caller from registration for the token id
+        registration_.write(token_id, 0);
```

Comment:
I am not sure, but I think the "good practice" does advise to ensure the "caller" is the owner of the token_id before removing it from the `registration_` liste ?

Reply:
Not according to the [Check - Effect - Interaction design pattern](https://fravoll.github.io/solidity-patterns/checks_effects_interactions.html) I think.

Reply:
My point was the same a earlier (but as you said, the check is done into the function `IERC721.transferFrom`), so this comment is not relevant anymore. 
Thank for the Check effect Interaction patern !